lib/Provider/Command/MessageSend.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCA\Mail\Provider\Command;
+
+use OCA\Mail\AppInfo\Application;
+use OCA\Mail\Db\LocalMessage;
+use OCA\Mail\Service\AccountService;
+use OCA\Mail\Service\Attachment\AttachmentService;
+use OCA\Mail\Service\OutboxService;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IConfig;
+use OCP\Mail\Provider\Exception\SendException;
+use OCP\Mail\Provider\IAddress;
+use OCP\Mail\Provider\IMessage;
+
+class MessageSend {
+
+	public function __construct(
+		protected IConfig $config,
+		protected ITimeFactory $time,
+		protected AccountService $accountService,
+		protected OutboxService $outboxService,
+		protected AttachmentService $attachmentService
+	) {
+	}
+
+	/**
+	 * Performs send operation
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 * @param string $serviceId mail account id
+	 * @param IMessage $message mail message object with all required parameters to send a message
+	 * @param array $options array of options reserved for future use
+	 *
+	 * @return LocalMessage
+	 *
+	 * @throws SendException on failure, check message for reason
+	 *
+	 */
+	public function perform(string $userId, string $serviceId, IMessage $message, array $options = []): LocalMessage {
+		// find user mail account details
+		try {
+			$account = $this->accountService->find($userId, (int)$serviceId);
+		} catch (\Throwable $e) {
+			throw new SendException('Error: occurred while retrieving mail account details', 0, $e);
+		}
+		// convert mail provider message to mail app message
+		$localMessage = new LocalMessage();
+		$localMessage->setType($localMessage::TYPE_OUTGOING);
+		$localMessage->setAccountId($account->getId());
+		$localMessage->setSubject($message->getSubject());
+		$localMessage->setBody($message->getBody());
+		// disabled due to issues caused by opening these messages in gui
+		//$localMessage->setEditorBody($message->getBody());
+		$localMessage->setHtml(true);
+		$localMessage->setSendAt($this->time->getTime());
+		// convert all mail provider attachments to local attachments
+		$attachments = [];
+		if (count($message->getAttachments()) > 0) {
+			// iterate attachments and save them
+			foreach ($message->getAttachments() as $entry) {
+				// determine if required parameters are set
+				if (empty($entry->getName()) || empty($entry->getType()) || empty($entry->getContents())) {
+					throw new SendException('Invalid Attachment Parameter: MUST contain values for Name, Type and Contents');
+				}
+				// convert mail provider attachment to mail app attachment
+				try {
+					$attachments[] = $this->attachmentService->addFileFromString(
+						$userId,
+						$entry->getName(),
+						$entry->getType(),
+						$entry->getContents()
+					)->jsonSerialize();
+				} catch (\Throwable $e) {
```

Comment:
```suggestion
				} catch (UploadException $e) {
```

Reply:
Wouldn't it be better to catch any and all errors and log them? That way the consuming app/service does not have to handle any/all the different errors that might occur. For instance, the addFileFromString function currently does not thrown a proper UploadException for any database errors, so if the filename is too long or the mime type is too long, this would cause a different error.